### PR TITLE
Implement FSharpQuickInfoProvider (by @OmarTawfik)

### PR DESCRIPTION
**Pull request by vasily-kirichenko**

### CODE REVIEW

File: vsintegration/src/FSharp.Editor/QuickInfoProvider.fs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                | Some(options) ->
+                    let! sourceText = document.GetTextAsync(cancellationToken) |> Async.AwaitTask
+                    let textSpan = TextSpan(0, sourceText.Length)
+                    let defines = CompilerEnvironment.GetCompilationDefinesForEditing(document.Name, options.OtherOptions |> Seq.toList)
+                    let tokens = FSharpColorizationService.GetColorizationData(document.Id, sourceText, textSpan, Some(document.FilePath), defines, cancellationToken)
+
+                    match tokens |> Seq.tryFind(fun t -> t.TextSpan.Contains(position)) with
+                    | Some(token) ->
+                        match token.ClassificationType with
+                        | ClassificationTypeNames.Identifier ->
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: In the GotoDefinition implementation we now have 

```
                match classifiedSpan.ClassificationType with
                | ClassificationTypeNames.ClassName
                | ClassificationTypeNames.DelegateName
                | ClassificationTypeNames.EnumName
                | ClassificationTypeNames.InterfaceName
                | ClassificationTypeNames.ModuleName
                | ClassificationTypeNames.StructName
                | ClassificationTypeNames.TypeParameterName
                | ClassificationTypeNames.Identifier -> 
```
So I'm assuming quickinfo won't work at operators with this PR.

I think you should abstract out the ``tryClassifyAtPosition`` implementation and use it in both places.

File: vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj [hunk 2]
```diff
@@ -93,6 +96,8 @@
   </ItemGroup>
   <ItemGroup>
     <Reference Include="mscorlib" />
+    <Reference Include="PresentationFramework" />
+    <Reference Include="WindowsBase" />
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj]: Why are these needed? thx

**Comment by dsyme** [comment]
Could you add one or two unit tests that exercises ``ProvideQuickInfo`` please?  There are many other unit tests already for the underlying logic of course, but having something in place to enable further future testing at the level of ``ProvideQuickInfo`` would be great

**Comment by dsyme** [comment]
@vasily-kirichenko Absolutely fantastic to see this in place.

File: vsintegration/src/FSharp.Editor/QuickInfoProvider.fs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    interface IDeferredQuickInfoContent with
+        override this.Create() : FrameworkElement =
+            let label = new Label()
+            label.Content <- content
+            label.Foreground <- SolidColorBrush(Colors.White)
+            upcast label
+
+[<Shared>]
+[<ExportQuickInfoProvider(PredefinedQuickInfoProviderNames.Semantic, FSharpCommonConstants.FSharpLanguageName)>]
+type internal FSharpQuickInfoProvider [<ImportingConstructor>] (serviceProvider: SVsServiceProvider) =
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: The C# and VB implementations have very different constructor signatures?

http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.EditorFeatures/QuickInfo/SemanticQuickInfoProvider.cs,13

Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: Please note I understand nothing about MEF composition

File: vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj (hunk 2 above)
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj]: PresentationFrameword is needed for `Label` type.
WindowsBase is for `FrameworkElement`.

File: vsintegration/src/FSharp.Editor/QuickInfoProvider.fs (hunk 1 above)
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: Done.

File: vsintegration/src/FSharp.Editor/QuickInfoProvider.fs (hunk 3 above)
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: Yes, C# provider imports other dependencies, but I don't think it's the problem. We successfully import `SVsServiceProvider` in several places. I don't know much about MEF either :(

**Comment by vasily-kirichenko** [comment]
> Could you add one or two unit tests that exercises ProvideQuickInfo please? There are many other unit tests already for the underlying logic of course, but having something in place to enable further future testing at the level of ProvideQuickInfo would be great

I'd like to make it work first. I still throws exception.

**Comment by vasily-kirichenko** [comment]
image

File: vsintegration/src/FSharp.Editor/QuickInfoProvider.fs [hunk 4]
```diff
@@ -51,7 +51,8 @@ type internal FSharpDeferredQuickInfoContent(content: string) =
 
 [<Shared>]
 [<ExportQuickInfoProvider(PredefinedQuickInfoProviderNames.Semantic, FSharpCommonConstants.FSharpLanguageName)>]
-type internal FSharpQuickInfoProvider [<ImportingConstructor>] (serviceProvider: SVsServiceProvider) =
+type internal FSharpQuickInfoProvider [<System.ComponentModel.Composition.ImportingConstructor>] 
+    ([<System.ComponentModel.Composition.Import(typeof<SVsServiceProvider>)>] serviceProvider: IServiceProvider) =
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: Yes, it works now.

**Comment by vasily-kirichenko** [comment]
It kind of works now. The tooltip is always positioned below all visible text, not close to the symbol.

**Comment by dsyme** [comment]
@vasily-kirichenko Yeah, getting the positioning right is really important.

**Comment by dsyme** [comment]
Great to see the progress!

File: vsintegration/src/FSharp.Editor/CommonHelpers.fs [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        (
+            documentKey: DocumentId, 
+            sourceText: SourceText, 
+            textSpan: TextSpan, 
+            fileName: string option, 
+            defines: string list, 
+            cancellationToken: CancellationToken
+        ) : List<ClassifiedSpan> =
+        try
+            let sourceTokenizer = tokenizers.GetValue ((defines, fileName), fun _ -> FSharpSourceTokenizer(defines, fileName))
```
Comment (vladima) [inline vsintegration/src/FSharp.Editor/CommonHelpers.fs]: I don't think this cache works given that CWT referential identity of an object as a key

Comment (dsyme) [inline vsintegration/src/FSharp.Editor/CommonHelpers.fs]: Ah ok. That was my code. Needs to be fixed

**Comment by dsyme** [comment]
@vasily-kirichenko See https://github.com/Microsoft/visualfsharp/pull/1855 for the methodology to add tests

File: vsintegration/src/FSharp.Editor/CommonHelpers.fs (hunk 5 above)
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/CommonHelpers.fs]: @vasily-kirichenko Could you check the identity used by ``ConditionalWeakTable`` and see if it uses ``.Equals`` and ``.GetHashCode`` or just object identity? Probably the latter.  Thanks.  (Not sure if there's any good key available in that case, maybe just remove this CWT for now)

File: vsintegration/src/FSharp.Editor/CompletionProvider.fs [hunk 6]
```diff
@@ -43,7 +43,13 @@ type internal FSharpCompletionProvider(workspace: Workspace, serviceProvider: SV
     let xmlMemberIndexService = serviceProvider.GetService(typeof<IVsXMLMemberIndexService>) :?> IVsXMLMemberIndexService
     let documentationBuilder = XmlDocumentation.CreateDocumentationBuilder(xmlMemberIndexService, serviceProvider.DTE)
 
-    static member ShouldTriggerCompletionAux(sourceText: SourceText, caretPosition: int, trigger: CompletionTriggerKind, getInfo: (unit -> DocumentId * string * string list)) =
+    static member ShouldTriggerCompletionAux
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/CompletionProvider.fs]: Please just use single line formatting for this.

File: vsintegration/src/FSharp.Editor/CommonHelpers.fs [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.Classification
+open Microsoft.CodeAnalysis.Text
+
+open Microsoft.VisualStudio.FSharp.LanguageService
+open Microsoft.FSharp.Compiler.SourceCodeServices
+
+module CommonHelpers =
+    type private SourceLineData
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/CommonHelpers.fs]: Please use single line formatting for this - multi-line is not useful - it's basically a record in any case

File: vsintegration/src/FSharp.Editor/CommonHelpers.fs [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        | FSharpTokenColorKind.UpperIdentifier -> ClassificationTypeNames.Identifier
+        | FSharpTokenColorKind.Number -> ClassificationTypeNames.NumericLiteral
+        | FSharpTokenColorKind.InactiveCode -> ClassificationTypeNames.ExcludedCode 
+        | FSharpTokenColorKind.PreprocessorKeyword -> ClassificationTypeNames.PreprocessorKeyword 
+        | FSharpTokenColorKind.Operator -> ClassificationTypeNames.Operator
+        | FSharpTokenColorKind.TypeName  -> ClassificationTypeNames.ClassName
+        | FSharpTokenColorKind.Default 
+        | _ -> ClassificationTypeNames.Text
+
+    let private scanSourceLine
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/CommonHelpers.fs]: Likewise - single line please

File: vsintegration/src/FSharp.Editor/QuickInfoProvider.fs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+// to allow for a more flexible syntax for defining the content of the tooltip.
+// The below interface should be discarded then or updated accourdingly.
+// https://github.com/dotnet/roslyn/pull/13623
+type internal FSharpDeferredQuickInfoContent(content: string) =
+    interface IDeferredQuickInfoContent with
+        override this.Create() : FrameworkElement =
+            let label = new Label()
+            label.Content <- content
+            label.Foreground <- SolidColorBrush(Colors.Black)
+            upcast label
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: Please use ``new Label(Content=content, Foreground=SolidColorBrush(Colors.Black))``

File: vsintegration/src/FSharp.Editor/QuickInfoProvider.fs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+[<Shared>]
+[<ExportQuickInfoProvider(PredefinedQuickInfoProviderNames.Semantic, FSharpCommonConstants.FSharpLanguageName)>]
+type internal FSharpQuickInfoProvider [<System.ComponentModel.Composition.ImportingConstructor>] 
+    ([<System.ComponentModel.Composition.Import(typeof<SVsServiceProvider>)>] serviceProvider: IServiceProvider) =
+
+    let xmlMemberIndexService = serviceProvider.GetService(typeof<SVsXMLMemberIndexService>) :?> IVsXMLMemberIndexService
+    let documentationBuilder = XmlDocumentation.CreateDocumentationBuilder(xmlMemberIndexService, serviceProvider.DTE)
+
+    static member ProvideQuickInfo
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: Single line signature please

File: vsintegration/src/FSharp.Editor/QuickInfoProvider.fs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+            document: Document, 
+            sourceText: SourceText, 
+            position: int, 
+            options: FSharpProjectOptions, 
+            filePath: string, 
+            textVersionHash: int,
+            cancellationToken: CancellationToken
+        ) = 
+        async {
+            let! parseResults = FSharpChecker.Instance.ParseFileInProject(filePath, sourceText.ToString(), options)
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: Please use ``ParseAndCheckFileInProject`` (see #1855 )

File: vsintegration/src/FSharp.Editor/QuickInfoProvider.fs [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+            let quickParseInfo = 
+                match tryClassifyAtPosition position with 
+                | None when textLineColumn > 0 -> tryClassifyAtPosition (position - 1) 
+                | res -> res
+
+            match quickParseInfo with 
+            | Some (islandColumn, qualifiers) -> 
+                //let tokenTag = QuickParse.CorrectIdentifierToken "" tokenTag
+                let! res = checkFileResults.GetToolTipTextAlternate(
+                               textLineNumber, islandColumn, textLine.ToString(), qualifiers, 189)
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: What is 189?  The tokenTag I guess. I think you can use ``tagOfToken`` on an identifier

File: vsintegration/src/FSharp.Editor/QuickInfoProvider.fs [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+                    | Some(_column, _island) ->
+                        let! textVersion = document.GetTextVersionAsync(cancellationToken) |> Async.AwaitTask
+                        let! toolTipElement = 
+                            FSharpQuickInfoProvider.ProvideQuickInfo(document, sourceText, position, options, document.FilePath, 
+                                textVersion.GetHashCode(), cancellationToken)
+                        match toolTipElement with
+                        | Some toolTipElement ->
+                            let dataTipText = XmlDocumentation.BuildDataTipText(documentationBuilder, toolTipElement) 
+                            return QuickInfoItem(textSpan, FSharpDeferredQuickInfoContent(dataTipText))
+                        | None -> return Unchecked.defaultof<_>
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: Please just use ``null`` given this is implementing a C# interface

Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfoProvider.fs]: (Or use ``None`` followed by ``Option.toObj`` at the end)

File: vsintegration/src/FSharp.Editor/GoToDefinitionService.fs [hunk 14]
```diff
@@ -43,70 +42,45 @@ type internal FSharpNavigableItem(document: Document, textSpan: TextSpan) =
...
-                | FSharpCheckFileAnswer.Succeeded(results) -> results
-
-            let! declarations = checkFileResults.GetDeclarationLocationAlternate (fcsTextLineNumber, islandColumn, textLine.ToString(), qualifiers, false)
-
-            match declarations with
-            | FSharpFindDeclResult.DeclFound(range) -> return Some(range)
-            | _ -> return None
-        | None -> return None
-    }
+    static member FindDefinition
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/GoToDefinitionService.fs]: Single line signature please

Comment (dsyme) [inline vsintegration/src/FSharp.Editor/GoToDefinitionService.fs]: (or formatted like the original)

Comment (dsyme) [inline vsintegration/src/FSharp.Editor/GoToDefinitionService.fs]: Maybe minimize the diff in this method?  Many of the changes are formatting

File: vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj (hunk 2 above)
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj]: ok

**Comment by dsyme** [comment]
@vasily-kirichenko This looks great.  I left a bunch of fairly minor comments.

File: src/fsharp/FSharp.Compiler/FSharp.Compiler.fsproj [hunk 15]
```diff
@@ -516,17 +516,6 @@
     <Reference Include="System.Reflection.Metadata"><HintPath>..\..\..\packages\System.Reflection.Metadata.1.4.1-beta-24227-04\lib\portable-net45+win8\System.Reflection.Metadata.dll</HintPath></Reference>
     <Reference Include="System.Collections.Immutable"><HintPath>..\..\..\packages\System.Collections.Immutable.1.2.0\lib\portable-net45+win8+wp8+wpa81\System.Collections.Immutable.dll</HintPath></Reference>
     <Reference Include="System.ValueTuple"><HintPath>..\..\..\packages\System.ValueTuple.4.3.0\lib\netstandard1.0\System.ValueTuple.dll</HintPath><Private>true</Private></Reference>
-    <BuiltProjectOutputGroupKeyOutput Include="$(FSharpSourcesRoot)\..\$(Configuration)\net40\bin\Microsoft.Build.Framework.dll" />
```
Comment (vasily-kirichenko) [inline src/fsharp/FSharp.Compiler/FSharp.Compiler.fsproj]: These lines prevents the project to be load in VS.

File: vsintegration/src/FSharp.Editor/CompletionProvider.fs (hunk 6 above)
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/CompletionProvider.fs]: I'll do, but I don't see your point. Is this better?

image

I cannot even see how many parameters this method has and what's the result type.

**Comment by vasily-kirichenko** [comment]
The position is fixed:

image

**Comment by vasily-kirichenko** [comment]
Added a test.

**Comment by vasily-kirichenko** [comment]
The only issue left: it does not respect theme colors.

**Comment by vasily-kirichenko** [comment]
I've ported somewhat hacky code from VFPT to get current theme colors (https://github.com/fsprojects/VisualFSharpPowerTools/blob/03c613acd3aa67cb5e3a339d3dcf1c7c783f30c2/src/FSharp.Editing.VisualStudio/Common/VSColors.fs)

It works:

image

image

If anybody knows a better way to do it, please tell us.

**Comment by forki** [comment]
Awesome!

File: src/fsharp/FSharp.Compiler/FSharp.Compiler.fsproj (hunk 15 above)
Comment (dsyme) [inline src/fsharp/FSharp.Compiler/FSharp.Compiler.fsproj]: Yep, I know.  But we shouldn't check in their removal until we find out what they are for.  People working on master who want to load FSHarp.Compiler into VS can just remove them by hand until tomorrow, thanks

**Comment by dsyme** [comment]
This is fantastic. Looks ready to go to me, once green

**Comment by dsyme** [comment]
@vasily-kirichenko Thank you!

**Comment by dsyme** [comment]
> Looks like it's pure hash table https://referencesource.microsoft.com/#mscorlib/system/runtime/compilerservices/ConditionalWeakTable.cs,464 Is it ok or I should remove the cache?

Yes, that's an object identity hash, not a call to obj.GetHashCode().  We should remove the cache for now

**Comment by dsyme** [comment]
Just to mention that after this change both the QUickInfo and errors show when hovering over a red squiggly

This is good, but disambiguating between these using colors will be necessary
capture7

**Comment by vasily-kirichenko** [comment]
@dsyme we don't have control over errors formatting.

**Comment by cloudRoutine** [comment]
> both the QuickInfo and errors show when hovering over a red squiggly

A long needed and overdue feature 😄

File: vsintegration/src/FSharp.Editor/CommonHelpers.fs [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+            | ClassificationTypeNames.Operator ->
+                let islandColumn = sourceText.Lines.GetLinePositionSpan(classifiedSpan.TextSpan).End.Character
+                Some (islandColumn, [""], classifiedSpan.TextSpan) 
+            | _ -> None
+        | _ -> None
+
+open System.Reflection
+open Microsoft.VisualStudio.Shell
+open System.Windows.Controls
+// A port of https://github.com/tomasr/viasfora/blob/master/Viasfora/Util/VsColors.cs
```
Comment (JoshVarty) [inline vsintegration/src/FSharp.Editor/CommonHelpers.fs]: If you guys are supporting VS 2010 in this release then you'll need to interact with colors this way. In which case I'd add a comment that says something like "We want to support Visual Studio 2010, but EnvironmentColors weren't added until Visual Studio 2012 so we have to load the VS2012 DLL dynamically and interact with themes using reflection."

If this .vsix isn't going to be used in VS2010 then I believe you can take a dependency on `Microsoft.VisualStudio.Shell.11.0` and refer to `EnvironmentColors` directly.

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/CommonHelpers.fs]: 1. We don't support _any_ VS version prior to 2017.
2. I've already removed that code.
3. We now use `IClassificationFormatMapService` https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/QuickInfoProvider.fs#L63

Comment (JoshVarty) [inline vsintegration/src/FSharp.Editor/CommonHelpers.fs]: 👍

**Comment by dsyme** [comment]
@vasily-kirichenko I'm noticing a **lot** of issues where VIsual Studio seems to hang while getting QuickInfo during debugging of F# code. 

I'm not sure if this is our problem, but it feels like it is.  It may make sense to work out how to turn off semantic/error quick info while debugging F# code.  At least we need to investigate further

**Comment by Pilchie** [comment]
Is it quick info, or determining data-tip spans?  Perhaps something in that code path is sub-optimal?

**Comment by dsyme** [comment]
@Pilchie Hard to tell. It's an outright hang of the VS IDE.  I got one stack trace where the main UI thread was on roslyn code calling through to a quick info provider in a blocking wait (the word "QuickInfo" was definitely on the stack.  I presume it was the F# provider for quick info tips. I'll try to repro and get better stack traces.

Is there a way to disable the F# Quick Info service when debugging is active if we need it?

thanks

**Comment by vasily-kirichenko** [comment]
Maybe it's not related, but I think this code https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/QuickInfoProvider.fs#L81-L88 is too heavy weight to get qualified name island, start column of the last part of it and its `TextSpan`. Maybe it's possible to use something like `QuickParse.GetCompleteIdentifierIsland` and create a `TextSpan` from it somehow?

**Comment by dsyme** [comment]
@vasily-kirichenko Hmmm... it felt more async-lock-up-the-system than that

**Comment by vasily-kirichenko** [comment]
however, code is not changed during debugging, so FCS should cache everything and answer very fast. It's strange. I've not able to investigate it today, been trying to build behind the proxy.

**Comment by vasily-kirichenko** [comment]
@dsyme on a hello world console application debugger works OK for me: it does not show tooltips and it does show proper debug information (current variable values).

On VFT solution it does not work, hanging on this this dialog:

image

Did you mean this behavior?

**Comment by cartermp** [comment]
@vasily-kirichenko I've noticed that too

**Comment by Pilchie** [comment]
If it hangs forever, a dump with heap would really make investigations go a lot faster.

**Comment by vasily-kirichenko** [comment]
@Pilchie it does not hang forever, sorry for misinformation. The dialog is shown until you move mouse somewhere else, after that VS is responsive until you hover the mouse over another symbol again.

**Comment by vasily-kirichenko** [comment]
I debugged it and I was wrong: `IQuickInfoProvider.GetItemAsync` is called in debug mode.

**Comment by vasily-kirichenko** [comment]
I make `IBreakpointResolutionService ` no-op and it has not helped:

```fsharp
interface IBreakpointResolutionService with
        member this.ResolveBreakpointAsync(_document: Document, _textSpan: TextSpan, _cancellationToken: CancellationToken): Task<BreakpointResolutionResult> =
            Task.FromResult null
```
(I mean "Getting datatip text" dialog still hangs)

**Comment by vasily-kirichenko** [comment]
I also make `IQuickInfoProvider ` noop, too and it has not helped either:

```fsharp
interface IQuickInfoProvider with
        override this.GetItemAsync(_document: Document, _position: int, _cancellationToken: CancellationToken): Task<QuickInfoItem> =
            Task.FromResult null
```

**Comment by vasily-kirichenko** [comment]
Obviously it hangs in `ILanguageDebugInfoService.GetDataTipInfoAsync`. I added `try...with` block around it and it did not help.

**Comment by dsyme** [comment]
@vasily-kirichenko COuld you add a new bug with repro steps you are using?  This seems important to fix (even though we're only seeing it in the VisualFSharp.sln right now, I presume it happens on other large solutions)

**Comment by vasily-kirichenko** [comment]
@dsyme done https://github.com/Microsoft/visualfsharp/issues/1915
